contracts/FlowStakingCollection.cdc

```diff
@@ -0,0 +1,844 @@
+
+                assert(
+                    amount <= unlockedBalance,
+                    message: "Insufficient total Flow balance"
+                )
+
+                self.unlockedTokensUsed = self.unlockedTokensUsed + amount
+
+                return <-self.unlockedVault.borrow()!.withdraw(amount: amount)
+            }
+        }
+
+        /// Deposits tokens back to a vault after being withdrawn from a Stake or Delegation.
+        /// Deposits to unlocked tokens first, if possible, followed by locked tokens
+        access(self) fun depositTokens(from: @FungibleToken.Vault) {
+            pre {
+                // This error should never be triggered in production becasue the tokens used fields
+                // should be properly managed by all the other functions
+                from.balance <= self.unlockedTokensUsed + self.lockedTokensUsed: "Cannot deposit more than is already used"
+            }
+
+            /// If there is a locked account, get the locked vault holder for depositing
+            if self.lockedVault != nil {
+  
+                if (from.balance <= self.unlockedTokensUsed) {
+                    self.unlockedTokensUsed = self.unlockedTokensUsed - from.balance
+
+                    self.unlockedVault.borrow()!.deposit(from: <-from)
+                } else {
+                    // Return unlocked tokens first
+                    self.unlockedVault.borrow()!.deposit(from: <-from.withdraw(amount: self.unlockedTokensUsed))
+                    self.unlockedTokensUsed = 0.0
+
+                    self.lockedTokensUsed = self.lockedTokensUsed - from.balance
+                    // followed by returning the difference as locked tokens
+                    self.lockedVault!.borrow()!.deposit(from: <-from)
+                }
+            } else {
+                self.unlockedTokensUsed = self.unlockedTokensUsed - from.balance
+                
+                // If there is no locked account, get the users vault capability and deposit tokens to it.
+                self.unlockedVault.borrow()!.deposit(from: <-from)
+            }
+        }
+
+        /// Returns true if a Stake or Delegation record exists in the StakingCollection for a given nodeID and optional delegatorID, otherwise false.
+        pub fun doesStakeExist(nodeID: String, delegatorID: UInt32?): Bool {
+            var tokenHolderNodeID: String? = nil
+            var tokenHolderDelegatorNodeID: String? = nil
+            var tokenHolderDelegatorID: UInt32?  = nil
+
+            // If there is a locked account, get the staking info from that account
+            if self.tokenHolder != nil {
+                if let _tokenHolder = self.tokenHolder!.borrow() {
+                    tokenHolderNodeID = _tokenHolder!.getNodeID()
+                    tokenHolderDelegatorNodeID = _tokenHolder!.getDelegatorNodeID()
+                    tokenHolderDelegatorID = _tokenHolder!.getDelegatorID()
```

Comment:
Due to the optional binding (if-let), `_tokenHolder` should be non-optional here:

```suggestion
                    tokenHolderNodeID = _tokenHolder.getNodeID()
                    tokenHolderDelegatorNodeID = _tokenHolder.getDelegatorNodeID()
                    tokenHolderDelegatorID = _tokenHolder.getDelegatorID()
```